Fix line item meta data overwrite

JustShah opened the pull request:
- Updated the logic to handle `line_item_identifier` in customer metadata.
- Ensured that `line_item_identifier` is stored as an array in the line item's customer metadata.
- Removed the generation of unique keys for `line_item_identifier`.

This change ensures that the `line_item_identifier` is correctly stored as an array in the line item's customer metadata, allowing multiple serial numbers to be associated with a single line item.

![image](https://github.com/user-attachments/assets/646dce78-5292-416e-af2a-6e04a18fd379)
![image](https://github.com/user-attachments/assets/57d9dc7b-f6a3-4f22-a1f8-200b8a0c5695)

JustShah committed:
Fix handling of line_item_identifier in customer metadata for line items

- Updated the logic to handle `line_item_identifier` in customer metadata.
- Ensured that `line_item_identifier` is stored as an array in the line item's customer metadata.
- Removed the generation of unique keys for `line_item_identifier`.

This change ensures that the `line_item_identifier` is correctly stored as an
array in the line item's customer metadata, allowing multiple serial numbers to be associated with a single line item.

fthobe commented on app/decorators/controllers/cart_line_items_controller_decorator.rb:
```diff
@@ -1,48 +1,46 @@
-            end
-          rescue ActiveRecord::RecordInvalid => e
-            @order.errors.add(:base, e.record.errors.full_messages.join(", "))
+    customer_metadata = params.require(:customer_metadata).permit! if params[:customer_metadata].present?
```
We have decided and documented how meta data should work. 
We had the discussion that sanitisation and conformity including checks as in: 
- do I overwrite metadata
- should I check what meta data is present
is the job of the interacting party and should therfor go into the frontend and not into the core. 

In this case the interaction is happening in the starter frontend, validation if identical line_item(s) are present should be done on the data entry section (logic powering the form) and not inside the core.

fthobe replied:
@JustShah please update on this

JustShah replied:
The changes were made in the Starter Frontend, not in the Core.

fthobe commented on app/decorators/controllers/cart_line_items_controller_decorator.rb:
```diff
@@ -1,48 +1,46 @@
+        @line_item = @order.contents.add(variant, quantity)
+        if customer_metadata.present?
+          customer_metadata.each do |key, value|
+            next unless key == "user_serial_number"
```
The logic should be generic.
Is meta data present > are key value pairs identical > build the array
It might be a consideration to pass all line item meta data as array even if only one value is present.

fthobe replied:
@JustShah please update on this

JustShah replied:
For `user_serial_number` we converted it into an array, even for a single value.

fthobe replied:
@JustShah As bespoken please name it line_item_identifier as we will correct incorrect data from the erp, this value will not be present on admin data, in addition, please consider that customer_metadata { user_serial_number

is redundant as customer and user are redundant.

JustShah replied:
1) Rename the customer metadata key to `line_item_identifier`
2) Rename the product column and name it as `Require Line Item Identifier`

![image](https://github.com/user-attachments/assets/46bcb19b-6834-4990-940f-f777fbd2dc02)

JustShah replied:
Resolved, Renamed key and column.

fthobe commented:
Some general notes: 
customer meta is customer data, it's our responsibility to check or validate as long as we don't provide the visualization and data entry layer. The logic is correct, though it should not alter the core.

fthobe reviewed (changes requested):
See comments

fthobe commented on app/decorators/controllers/cart_line_items_controller_decorator.rb:
```diff
@@ -1,48 +1,46 @@
+    quantity = params[:quantity].present? ? params[:quantity].to_i : 1
 
-        customer_metadata = params.require(:customer_metadata).permit! if params[:customer_metadata].present?
-        # 2,147,483,647 is crazy. See issue https://github.com/spree/spree/issues/2695.
```
I think this comment was incorrectly moved. Restore it at the correct location.

fthobe replied:
@JustShah please update on this

JustShah replied:
I haven't made any changes to the comment, I just corrected the indentation.

fthobe commented:
We need to decide where to put this. 
@JustShah we talk tonight about this one.

JustShah committed:
Rename require_identifier to require_line_item_identifier

- Update attribute and method names for consistency and clarity.
- Adjust views and migration to reflect the new naming convention.
